resource/config_map.go

```diff
@@ -0,0 +1,198 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type ConfigMapConfig struct {
+	// DesiredFn is function returning a desired ConfigMap object for the
+	// given custom resource object. Returned ConfigMap must have name and
+	// namespace equal to the values of Name and Namespace fields.
+	DesiredFn func(context.Context, interface{}) (*corev1.ConfigMap, error)
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+
+	// Name of the reconciled ConfigMap. It must be the same name as in
+	// ConfigMap returned by DesiredFn.
+	Name string
+	// Namespace of the reconciled ConfigMap. It must be the same namespace
+	// is in ConfigMap returned by DesiredFn.
+	Namespace string
```

Comment:
I tried to take name and namespace from the desired object in first prototype. I don't think it's a good idea, because sometimes computing desired state is impossible on deletion. We ran into this problem many times. That's why we have `key.IsDeleted`  functions. Having a separate name and namespace fields also improves log messages.

Reply:
I think we could just take the name and namespace from the desired state and on delete remove all resources we get from current state.

Reply:
Feels pretty constructed. Lets maybe have the simple version and use it and once we get such a problem see what the problem is and how we can solve it.

Reply:
For all kinds of unexpected return arguments we can just put validation in place. All good.